### PR TITLE
chore: migrate base package from axios to fetch/undici

### PR DESCRIPTION
**Pull request by ava-silver**
### What and why?

Replace `axios` (and its transitive dependency `follow-redirects`, which has had multiple CVEs) with `undici.fetch` (what powers native fetch in node) in `packages/base`. This is **Stage 1** of a multi-stage migration — it covers the base package only; downstream plugins continue to compile via backward-compat type aliases.

This is something we have been discussing internally and was also brought up in #2199

### How?

- New `request.ts` module wraps `undici.fetch` with an axios-like API surface (`httpRequest`, `RequestError`, `RequestConfig`, `RequestResponse`)
- Proxy support via `undici.ProxyAgent` (explicit proxy URL) and `EnvHttpProxyAgent` (env-var auto-detection)
- All `packages/base` consumers (`apikey`, `flare`, `trace`, `tag`, `measure`, `gitdb`) switched to new request module
- `interfaces.ts` exports backward-compat type aliases (`AxiosRequestConfig`, `AxiosPromise`) so downstream plugins compile unchanged
- Tests rewired from `jest.spyOn(axios)` to `jest.mock('../request')` pattern

**Dependency changes:**

- Removed: `axios` (+ transitive `follow-redirects`), `proxy` (test devDep)
- Added: `undici` 7.24.6 (zero transitive deps)

### Known Stage 1 workarounds (cleaned up in Stage 2)

- `RequestConfig` fields typed as `any` for compat with axios-typed plugin callers
- `RequestBuilder = (args: any) => Promise<any>` — permissive so both old and new callers compile
- `AxiosRequestConfig` / `AxiosPromise` type aliases exported from `interfaces.ts`
- `getAxiosError` test helper typed as `any` for synthetics test compat
- `proxy-agent` remains as a dependency (used by synthetics WebSocket tunnel and `datadog-metrics`)

### Behavioral note

`undici.ProxyAgent` only tunnels HTTPS traffic (via CONNECT); plain HTTP bypasses the proxy. The proxy routing tests were removed accordingly. This matches real-world usage since all Datadog API endpoints are HTTPS.

### Remaining stages

- **Stage 2**: Migrate 9 plugin packages — mechanical type swaps (`AxiosRequestConfig` → `RequestConfig`, etc.) across ~23 files, remove compat aliases
- **Stage 3**: Migrate `plugin-synthetics` separately — has custom retry logic, `paramsSerializer`, and heavier test mocking
- **Stage 4**: Remove `proxy-agent` dep, clean up lockfile, verify `follow-redirects` is fully gone from the tree

### Review checklist

- [x] Feature or bugfix MUST have appropriate tests (unit, integration)
- [x] All existing tests pass
- [x] CI still passes

### CODE REVIEW

**Comment by ava-silver** [comment]
* **#2249** <a href="https://app.graphite.com/github/pr/DataDog/datadog-ci/2249?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2248** <a href="https://app.graphite.com/github/pr/DataDog/datadog-ci/2248?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2247** <a href="https://app.graphite.com/github/pr/DataDog/datadog-ci/2247?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2201** <a href="https://app.graphite.com/github/pr/DataDog/datadog-ci/2201?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/DataDog/datadog-ci/2201?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Review by nina9753 — APPROVED** [review]

**Review by rodrigo-roca — APPROVED** [review]

File: packages/base/src/helpers/__tests__/testing-tools.ts [hunk 1]
```diff
@@ -137,10 +137,12 @@ export const createCommand = <T extends Command>(
...
+  new RequestError(
+    message ?? 'Request failed',
+    {baseURL: MOCK_BASE_URL, url: 'example'},
+    {data: {errors}, status, statusText: ''}
+  )
 
-  return serverError
-}
+// Backward-compat alias for plugin tests (removed in Stage 2)
+export const getAxiosError: any = getRequestError
```
Comment (GabrielAnca) [inline packages/base/src/helpers/__tests__/testing-tools.ts]: Can we use something more specific than `any` to ensure we don't break anything with this change? I'm just worried that setting `any` will make almost anything to compile after the change

File: packages/base/src/helpers/request.ts [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+import type {Readable} from 'stream'
+
+import {EnvHttpProxyAgent, ProxyAgent, fetch} from 'undici'
+
+export interface RequestConfig {
+  baseURL?: string
+  data?: any
+  dispatcher?: any // undici Dispatcher — typed as any to avoid @types/node version conflict
+  headers?: any // permissive to allow AxiosHeaders during Stage 1 migration
+  maxBodyLength?: number // accepted for compat, ignored (no limit with fetch)
```
Comment (GabrielAnca) [inline packages/base/src/helpers/request.ts]: Should we move this to the bottom together with the other compat fields?

File: packages/plugin-synthetics/src/__tests__/api.test.ts [hunk 3]
```diff
@@ -285,7 +305,7 @@ describe('dd-api', () => {
         baseV2Url: `http://127.0.0.1:${port}`,
       })
 
-      await expect(localApi.getSyntheticsOrgSettings).rejects.toThrow('socket hang up')
+      await expect(localApi.getSyntheticsOrgSettings).rejects.toThrow('other side closed')
```
Comment (GabrielAnca) [inline packages/plugin-synthetics/src/__tests__/api.test.ts:308]: Wondering if the change in error message could be a breaking change. May users rely on this exact error message? This looks like an edge case, but I'm just asking 😄

**Review by GabrielAnca — APPROVED** [review]
Looks great! Just some minor comments

**Review by rpelliard — APPROVED** [review]

File: packages/plugin-synthetics/src/__tests__/api.test.ts (hunk 3 above)
Comment (ava-silver) [reply packages/plugin-synthetics/src/__tests__/api.test.ts:308]: This string comes from the underlying transport layer and is never documented or surfaced in our CLI output as a meaningful user-visible message -- it only appears in raw error stacks, which to me feels more like an implementation detail change than a breaking change. There are potentially others that could arise depending on network conditions, and unfortunately there isnt an existing mapping we can use afaik

That being said, we can definitely document it in the release that this comes in, mentioning that users who are relying on axios specific errors or other axios specific implementation details should be aware before updating

File: packages/base/src/commands/trace/api.ts [hunk 4]
```diff
@@ -1,26 +1,20 @@
 import type {Payload} from './interfaces'
-import type {AxiosPromise, AxiosRequestConfig, AxiosResponse} from 'axios'
+import type {RequestBuilder} from '@datadog/datadog-ci-base/helpers/interfaces'
 
 import {getRequestBuilder} from '@datadog/datadog-ci-base/helpers/utils'
 
-// Dependency follows-redirects sets a default maxBodyLength of 10 MB https://github.com/follow-redirects/follow-redirects/blob/b774a77e582b97174813b3eaeb86931becba69db/index.js#L391
-// We don't want any hard limit enforced by the CLI, the backend will enforce a max size by returning 413 errors.
-const maxBodyLength = Infinity
```
Comment (Drarig29) [inline packages/base/src/commands/trace/api.ts:8]: Did you confirm fetch/undici doesn't enforce a max body length either?

File: packages/base/src/helpers/__tests__/utils.test.ts [hunk 5]
```diff
@@ -247,132 +255,38 @@ describe('utils', () => {
...
-          url: 'test-from-proxy',
-        })
-        expect(targetServer.spy.mock.calls.length).toBe(1)
-        expect(proxyServer.spy.mock.calls.length).toBe(1)
-      } finally {
-        await targetServer.close()
-        await proxyServer.close()
-      }
-    })
+    // Proxy routing for plain HTTP is not testable here without TLS certs — undici
```
Comment (Drarig29) [inline packages/base/src/helpers/__tests__/utils.test.ts]: that comment is weird with the em-dash 😁

File: packages/base/src/helpers/__tests__/utils.test.ts [hunk 6]
```diff
@@ -247,132 +255,38 @@ describe('utils', () => {
...
-        })
-        expect(targetServer.spy.mock.calls.length).toBe(1)
-        expect(proxyServer.spy.mock.calls.length).toBe(1)
-      } finally {
-        await targetServer.close()
-        await proxyServer.close()
-      }
-    })
-
-    test('Proxy configured through env var', async () => {
```
Comment (Drarig29) [inline packages/base/src/helpers/__tests__/utils.test.ts:332]: Do we still test somewhere that the proxy can be configured through env vars?

File: packages/plugin-synthetics/src/__tests__/api.test.ts (hunk 3 above)
Comment (Drarig29) [reply packages/plugin-synthetics/src/__tests__/api.test.ts:308]: sounds good! we can have a small note about it

**Review by Drarig29 — APPROVED** [review]

File: packages/base/src/helpers/__tests__/utils.test.ts (hunk 5 above)
Comment (ava-silver) [reply packages/base/src/helpers/__tests__/utils.test.ts]: i'll un-ai-ify it lol

File: packages/base/src/commands/trace/api.ts (hunk 4 above)
Comment (ava-silver) [reply packages/base/src/commands/trace/api.ts:8]: Yep, the docs don't mention anything about it and everything I've read says that undici uses automatic streaming to handle it so theres no need for the maxBodyLength there. if we were setting this setting to a specific limit, we couldn't set that the same way in undici to make it error (we'd have to do a check after the fact). The TLDR is that we're all good here

File: packages/base/src/helpers/__tests__/utils.test.ts (hunk 6 above)
Comment (ava-silver) [reply packages/base/src/helpers/__tests__/utils.test.ts:332]: The old test (`'Proxy configured through env var'`) worked by routing plain HTTP through a local proxy server. With undici, `ProxyAgent` only tunnels HTTPS (via CONNECT), so that approach no longer works.

It is technically testable -- you'd spin up a TCP server, set `HTTPS_PROXY=http://localhost:PORT`, make a request to an `https://` target, and assert the server received a `CONNECT` request. But it felt like a lot of plumbing just to test that undici's `EnvHttpProxyAgent` reads env vars correctly, which is really undici's own test coverage responsibility.

I've updated the existing test to assert the dispatcher is an `EnvHttpProxyAgent` instance (rather than just truthy), which at least verifies we're wiring up the right dispatcher type. Happy to add the full integration test if you think it's worth it, maybe in a follow-up PR if thats okay?

Comment (Drarig29) [reply packages/base/src/helpers/__tests__/utils.test.ts:332]: nah, no need it for it. indeed, it's undici's responsibility
